File: meeting-flutter/meeting_kit_bridge/lib/src/meeting_message_channel_service.dart
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import 'package:fluttermodule/meeting_kit_bridge.dart';
import 'package:fluttermodule/src/callback.dart';
import 'package:fluttermodule/src/service.dart';
import 'package:netease_meeting_kit/meeting_core.dart';

class MeetingMessageChannelServiceBridge extends Service
    implements NEMeetingMessageChannelListener {
  final MeetingKitBridge meetingKitBridge;

  MeetingMessageChannelServiceBridge.asService(this.meetingKitBridge) {
    meetingKitBridge.meetingMessageChannelService
        .addMeetingMessageChannelListener(this);
  }

  @override
  String get name => 'meetingMessageChannel';

  String mapMethodName(String method) => name + '.' + method;

  @override
  Future handleCall(String method, arguments) {
    switch (method) {
      case 'queryUnreadMessageList':
        return _handleQueryUnreadMessageList(arguments);
      case 'clearUnreadCount':
        return _handleClearUnreadCount(arguments);
      case 'deleteAllSessionMessage':
        return _handleDeleteAllSessionMessage(arguments);
      case 'getSessionMessagesHistory':
        return _handleGetSessionMessagesHistory(arguments);
    }
    return super.handleCall(method, arguments);
  }

  Future<Map> _handleQueryUnreadMessageList(arguments) async {
    assert(arguments['sessionId'] is String);
    return meetingKitBridge.meetingMessageChannelService
        .queryUnreadMessageList(
      arguments['sessionId'],
    )
        .then((value) {
      final data = value.data?.map((e) => e.toMap()).toList();
      return Callback.wrap(
        'queryUnreadMessageList',
        value.code,
        msg: value.msg,
        data: data,
      ).result;
    });
  }

  Future<Map> _handleClearUnreadCount(arguments) async {
    assert(arguments['sessionId'] is String);
    return meetingKitBridge.meetingMessageChannelService
        .clearUnreadCount(
      arguments['sessionId'],
    )
        .then((value) {
      return Callback.wrap('clearUnreadCount', value.code, msg: value.msg)
          .result;
    });
  }

  Future<Map> _handleDeleteAllSessionMessage(arguments) async {
    assert(arguments['sessionId'] is String);
    return meetingKitBridge.meetingMessageChannelService
        .deleteAllSessionMessage(
      arguments['sessionId'],
    )
        .then((value) {
      return Callback.wrap(
        'deleteAllSessionMessage',
        value.code,
        msg: value.msg,
      ).result;
    });
  }

  Future<Map> _handleGetSessionMessagesHistory(arguments) async {
    assert(arguments['param'] is Map);
    final param = NEMeetingGetMessageHistoryParams.fromMap(
        arguments['param'] as Map<String, dynamic>);
    return meetingKitBridge.meetingMessageChannelService
        .getSessionMessagesHistory(param)
        .then((value) {
      final data = value.data?.map((e) => e.toMap()).toList();
      return Callback.wrap(
        'getSessionMessagesHistory',
        value.code,
        msg: value.msg,
        data: data,
      ).result;
    });
  }

  @override
  void onSessionMessageAllDeleted(
      String sessionId, NEMeetingSessionTypeEnum sessionType) {
    meetingKitBridge.channel
        .invokeMethod(mapMethodName('notifyMeetingSessionMessageAllDeleted'), {
      'sessionId': sessionId,
      'sessionType': sessionType.value,
    });
  }

  @override
  void onSessionMessageDeleted(NEMeetingSessionMessage message) {
    meetingKitBridge.channel.invokeMethod(
      mapMethodName('notifyMeetingSessionMessageDeleted'),
      message.toMap(),
    );
  }

  @override
  void onSessionMessageReceived(NEMeetingSessionMessage message) {
    meetingKitBridge.channel.invokeMethod(
      mapMethodName('notifyMeetingSessionMessageReceived'),
      message.toMap(),
    );
  }

  @override
  void onSessionMessageRecentChanged(List<NEMeetingRecentSession> messages) {
    meetingKitBridge.channel.invokeMethod(
      mapMethodName('notifyMeetingSessionMessageRecentChanged'),
      [
        for (var message in messages) message.toMap(),
      ],
    );
  }
}

File: meeting-electron/packages/meeting-kit-core/src/kit/interface/service/meeting_message_channel_service.ts
import {
  NEMessageSearchOrder,
  NEResult,
  NERoomSessionTypeEnum,
} from 'neroom-types'

export type NEMeetingGetMessageHistoryParams = {
  /** 获取聊天对象的Id（好友帐号，群ID等） 会话Id */
  sessionId: string
  /** 查询开启时间点 */
  fromTime?: number
  /** 查询截止时间点 */
  toTime: number
  /** 条数限制 限制0~100，否则414。其中0会被转化为100 */
  limit: number
  /** 查询方向,默认从大到小排序 */
  searchOrder?: NEMessageSearchOrder
}

export type NEMeetingSessionTypeEnum = NERoomSessionTypeEnum

export type NEMeetingSessionMessage = {
  /** 会话 id */
  sessionId: string
  /** 会话类型 */
  sessionType: NEMeetingSessionTypeEnum
  /** 消息 id */
  messageId: string
  /** 消息内容 */
  data: string
  /** 消息时间 */
  time: number
}

export type NEMeetingRecentSession = {
  /**
   * 获取聊天对象的Id（好友帐号，群ID等）
   * 会话Id
   */
  sessionId: string
  /**
   * 获取与该联系人的最后一条消息的发送方的帐号
   * 发送者帐号
   */
  fromAccount: string
  /**
   * 获取与该联系人的最后一条消息的发送方的昵称
   * 发送者昵称
   */
  fromNick: string
  /**
   * 会话类型
   */
  sessionType: NEMeetingSessionTypeEnum
  /**
   * 最近一条消息的UUID
   */
  recentMessageId: string
  /**
   * 该联系人的未读消息条数
   */

  unreadCount: number
  /**
   * 最近一条消息的缩略内容
   */
  content: string
  /**
   * 最近一条消息的时间，单位为ms
   */
  time: number
}

export type NEMeetingMessageChannelListener = {
  /**
   * 接收到自定义消息时会回调该方法
   * @param message 自定义会话消息
   */
  onSessionMessageReceived?(message: NEMeetingSessionMessage): void
  /**
   * 会话消息未读数变更时会回调该方法
   *
   * @param messages 未读消息列表
   */
  onSessionMessageRecentChanged?(messages: NEMeetingRecentSession[]): void
  /**
   * 会话消息被删除时会回调该方法
   *
   * @param message 自定义会话消息
   */
  onSessionMessageDeleted?(message: NEMeetingSessionMessage): void
  /**
   * 会话消息全部被删除时会回调该方法
   *
   * @param sessionId 会话id
   * @param sessionType 会话类型
   */
  onSessionMessageAllDeleted?(
    sessionId: string,
    sessionType: NEMeetingSessionTypeEnum
  ): void
}

interface NEMeetingMessageChannelService {
  /**
   * 添加自定义消息监听
   * @param listener 消息监听器
   */
  addMeetingMessageChannelListener: (
    listener: NEMeetingMessageChannelListener
  ) => void
  /**
   * 移除自定义消息监听
   * @param listener 消息监听器
   */
  removeMeetingMessageChannelListener: (
    listener: NEMeetingMessageChannelListener
  ) => void
  /**
   * 获取指定会话的未读消息列表
   * @param sessionId 会话id
   */
  queryUnreadMessageList(
    sessionId: string
  ): Promise<NEResult<NEMeetingSessionMessage[]>>
  /**
   * 清除指定会话的未读消息数
   * @param sessionId 会话id
   */
  clearUnreadCount(sessionId: string): Promise<NEResult<void>>
  /**
   * 删除指定会话的所有消息
   * 该接口会触发，{@link NEMeetingMessageChannelListener.onSessionMessageAllDeleted} 回调通知
   * @param sessionId 会话id
   */
  deleteAllSessionMessage(sessionId: string): Promise<NEResult<void>>
  /**
   * 获取指定会话的历史消息
   * @param param 查询参数
   */
  getSessionMessagesHistory(
    param: NEMeetingGetMessageHistoryParams
  ): Promise<NEResult<NEMeetingSessionMessage[]>>
}

export default NEMeetingMessageChannelService
